### DISCUSSION TITLE
Set an environment variable depending on the last run endpoint

### DISCUSSION THREAD

**Discussion by jason2li** [Kong/insomnia#3829] [Ideas]
**What is your question?**
I'm using JWT authentication.  I have a /SignIn endpoint and a /Refresh endpoint.  Both of them return the same data model, which includes an access token.  I'd like to save the last called endpoint to an environment variable called "accessToken".  Is this possible?

**What is the context?**
In my scenario, the user signs in using the /SignIn endpoint.  I then save the response to an environment variable called "accessToken".  After a certain amount of time that access token will expire, so they need to refresh it by calling the /Refresh endpoint.  Because of this call, the original "accessToken" variable is no longer valid.  I would like to update the variable to use the new, valid, access token from the /Refresh endpoint.

My understanding of both "chaining" and "environment variables" is that they can only consume from a single endpoint.

Is there a way to do this?

**Comment by zappatero** [comment]
I would love to have this features as well. Like in Postman "setEnvironmentVariable".

**Comment by ksugiarto** [comment]
This is a nice and important feature 👍

**Comment by alexejk** [comment]
I was struggling with this for a while, but i realized that with help of a simple plugin I can make it with:

- Response hook to parse response of desired methods and store it in plugin storage context
- Template tag that is used on methods that need part of the response

While i think it would be nice to `setEnvironmentVariable` I believe it will be challenging as there are multiple layers involved (base, env, folder override). Plugin store works very well, and they are very easy to create.

**Comment by Eddie-Hartman** [comment]
@develohpanda 
Just reaching out to a contributor here. This feature has been requested multiple times and seems like some people are going to struggle (myself and team included) to migrate from Postman if we can't store responses in variables. #217 and #1167 along with this are asking for it.

It seems like the plugin system has a few plugins to allow this kind of functionality, but it's not ideal.

**Comment by develohpanda** [comment]
Thanks for tagging me @Eddie-Hartman, there's a lot of activity and unfortunately some things go under the radar. Tagging @wdawson here, Wils is our new PM for Insomnia and will be the one who can help prioritise.

As a previous comment mentions this is certainly an involved piece of work _but_ I can see a way in which it could work.

**Reply by wdawson** [reply]
@Eddie-Hartman and @develohpanda thanks for raising this! I've converted to a GitHub Discussion that will hopefully allow for better, well, discussion and visibility for the community too. Definitely something we'll look at but I can't promise anything on the horizon yet.

For now, any examples of how you'd plan to use this functionality would be greatly appreciated. You can start new threads below where we can discuss each. That will just help us build a broader understanding of the ask and make sure that we build the right thing when we come to it.

**Reply by bamhm182** [reply]
@jason2li did a good job of explaining my use case at the top of this discussion. I'm just looking for a way to have a global JWT access token that comes from other `/token/obtain` or `/token/refresh` depending on which one was run most recently.

**Reply by Eddie-Hartman** [reply]
This is a good example of one of our use cases as well.

**Reply by ThisNoName** [reply]
My service accepting tokens from 3 different providers, so chaining won't work.  What I did was

- On token call, use [save-variables](https://insomnia.rest/plugins/insomnia-plugin-save-variables) plugin to save the token
- Use [default-headers](https://insomnia.rest/plugins/insomnia-plugin-default-headers) plugin to set authorization header from the variable.
- On a few unsecure endpoints, use folder environment to remove header.  I have to customize the plugin a little for that.

All seem working well.  On a bigger picture, we could really use multi tab and no auto save.  Without those, it's hard to do ad-hoc testing consistently in a collection with dozens calls
